runtime/globroots.c

```diff
@@ -17,134 +17,196 @@
 
 /* Registration of global memory roots */
 
-#include "caml/memory.h"
-#include "caml/misc.h"
 #include "caml/mlvalues.h"
 #include "caml/roots.h"
 #include "caml/globroots.h"
-#include "caml/callback.h"
-#include "caml/platform.h"
-#include "caml/alloc.h"
-#include "caml/shared_heap.h"
-#ifdef NATIVE_CODE
+#include "caml/skiplist.h"
 #include "caml/stack.h"
-#endif
-
-/* A caml_root is in fact a value. We don't expose that fact outside
-   of this file so that C code doesn't attempt to directly modify it.
-   The value points to a block on the shared heap with the following
-   fields:
-
-    0: the actual root, as set by caml_modify_root
-    1: an integer flag stating whether this root has been deleted
-    2: the next root in roots_all
-
-   The roots are not scanned during minor GC. Instead, since the root
-   blocks are all on the shared heap, pointers from roots to a minor
-   heap will be detected using the normal inter-generational pointer
-   mechanism. */
 
 static caml_plat_mutex roots_mutex = CAML_PLAT_MUTEX_INITIALIZER;
-static value roots_all = Val_unit;
 
+/* legacy multicore API that we need to fix */
 CAMLexport caml_root caml_create_root(value init)
 {
   CAMLparam1(init);
-  CAMLlocal1(v);
-  v = caml_alloc_shr(3, 0);
-  caml_initialize_field(v, 0, init);
-  caml_initialize_field(v, 1, Val_int(1));
+  
+  value* v = (value*)malloc(sizeof(value));
```

Comment:
Do we want any error handling here?

Reply:
This is a good point. I've switched from `malloc` to `caml_stat_alloc` and I've changed the `_noexc` version to return NULL correctly.